### DISCUSSION TITLE
WinPR SSPI Implementation Question

### DISCUSSION THREAD

**Discussion by geiseri** [FreeRDP/FreeRDP#10437] [Q&A] [answered]
I am working on an implementation of the TSGU server side and I have some issues with the section 3.5.5 (Message Processing Events and Sequencing Rules) and the auth negotiate part.  The winpr SSPI implementation seems to handle the negotiate protocol enough that it can use the resulting hashes to get to `AcceptSecurityContext` twice.  Unfortunately it only gets that far before i get the error:

```
[ERROR][com.winpr.sspi.NTLM] - [ntlm_fetch_ntlm_v2_hash]: Error: Could not find user in SAM database
``` 

Unfortunately gdb is less than helpful because by the time I step down to that point the client has become bored with me and wanders off. 

I used `winpr-hash  -u #### -d #### -p #### -f sam -v 2` to generate the SAM entry and placed the file in `/etc/winpr/SAM` with no impact on the system.  Ideally I would like to use kerberos but it seems that during the `AcceptSecurityContext` it sends back `NTLM` instead of kerberos by default.  I figured NTLM was a walk before you crawl strategy for now.

Any hints on what I could be doing wrong?  I have been following the "Using SSPI with a Windows Sockets Server" example code since it seems to be what the webserver projects use for their SSPI auth modules.  There was some conversation from the client side, but most of them ended with "upgrading to freerdp 2/3 fixed it" ;)

**Comment by akallabeth** [comment, marked as answer]
you need to use the `NTLMv1` hash in the `SAM` file, that is then hashed again to gain the `NTLMv2` hashes to compare internally.

**Reply by geiseri** [reply]
That was the magic! One other thing to note (I am sure you know this already, so this is just for posterity) is that the name and domain need to match.  This is only an issue if you are mixing UPN with NETBIOS names.
